crates/cli-message/src/parse.rs

```diff
@@ -0,0 +1,228 @@
+    Code,
+    ExitCode,
+}
+
+impl TryFrom<proc_macro2::Ident> for SubAttr {
+    type Error = ParseMessageError;
+
+    fn try_from(ident: proc_macro2::Ident) -> Result<Self, Self::Error> {
+        match ident.to_string().as_str() {
+            "message" => Ok(SubAttr::Message),
+            "code" => Ok(SubAttr::Code),
+            "exit_code" => Ok(SubAttr::ExitCode),
+            _ => Err(ParseMessageError::UnknownIdent),
+        }
+    }
+}
+
+// Parses the token stream contained in wrapping parentheses of the message attr
+// ie #[message(<token-stream>)]
+pub fn parse_message_attribute(toks: TokenStream2) -> Result<ImMessage, ParseMessageError> {
+    println!("{:?}", toks.to_string());
+    let mut im = ImMessage {
+        content: None,
+        code: None,
+        exit_code: None,
+    };
+    let mut ident_being_parsed: Option<SubAttr> = None;
+
+    fn set_opt_once<T>(opt: &mut Option<T>, val: T) -> Result<(), ParseMessageError> {
+        if opt.is_some() {
+            return Err(ParseMessageError::FieldSetTwice);
+        }
+
+        *opt = Some(val);
+        Ok(())
+    }
+
+    for (i, tok) in toks.into_iter().enumerate() {
+        match tok {
+            TokenTree::Literal(lit) => {
+                let lit = lit.to_string().replace("\"", "");
+
+                // handle the default case where content is first arg
+                if i == 0 {
+                    im.content = Some(lit);
+                    ident_being_parsed = None;
+                    continue;
+                }
+
+                if let Some(ident) = ident_being_parsed {
+                    match ident {
+                        SubAttr::Message => set_opt_once(&mut im.content, lit)?,
+                        SubAttr::Code => set_opt_once(&mut im.code, lit)?,
+                        SubAttr::ExitCode => {
+                            let exit_code: i32 = lit
+                                .parse()
+                                .map_err(|e| ParseMessageError::ExitCodeParse(e))?;
+                            set_opt_once(&mut im.exit_code, exit_code)?;
+                        }
+                    }
+                } else {
+                    return Err(ParseMessageError::LiteralWithoutIdent);
+                }
+
+                ident_being_parsed = None;
+            }
+            TokenTree::Ident(ident) => {
+                ident_being_parsed = Some(SubAttr::try_from(ident)?);
+            }
+            _ => continue,
+        }
+    }
+
+    Ok(im)
+}
+
+pub fn derive_message_from_im(
+    im: ImMessage,
+    enum_name: String,
+    variant_name: String,
+) -> Result<Message, ParseMessageError> {
+    let content = match im.content {
+        Some(content) => content,
+        None => return Err(ParseMessageError::MissingMessageAttribute),
+    };
+
+    Ok(Message {
+        content,
+        code: im.code.unwrap_or_else(|| {
+            build_message_code_from_enum(enum_name.to_string(), variant_name.to_string())
```

Comment:
Both of those values are already strings
```suggestion
            build_message_code_from_enum(enum_name, variant_name)
```